lib/auth/grpcserver.go

```diff
@@ -1966,14 +1967,54 @@ func (g *GRPCServer) DeleteAllKubernetesServers(ctx context.Context, req *authpb
 
 // maybeDowngradeRole tests the client version passed through the gRPC metadata,
 // and if the client version is unknown or less than the minimum supported
-// version for some features of the role returns a shallow copy of the given
-// role downgraded for compatibility with the older version.
+// version, it returns a shallow copy of the given role downgraded
+// for compatibility with the older version.
 func maybeDowngradeRole(ctx context.Context, role *types.RoleV6) (*types.RoleV6, error) {
-	// Teleport 16 supports all role features that Teleport 15 does,
-	// so no downgrade is necessary.
+	clientVersionString, ok := metadata.ClientVersionFromContext(ctx)
+	if !ok {
+		// This client is not reporting its version via gRPC metadata. Teleport
+		// clients have been reporting their version for long enough that older
+		// clients won't even support v6 roles at all, so this is likely a
+		// third-party client, and we shouldn't assume that downgrading the role
+		// will do more good than harm.
+		return role, nil
+	}
+
+	clientVersion, err := semver.NewVersion(clientVersionString)
+	if err != nil {
+		return nil, trace.BadParameter("unrecognized client version: %s is not a valid semver", clientVersionString)
+	}
+
+	role = maybeDowngradeRoleVersionToV7(role, clientVersion)
 	return role, nil
 }
 
+var minSupportedRoleV8Version = semver.New(utils.VersionBeforeAlpha("17.0.0"))
+
+// maybeDowngradeRoleVersionToV7 downgrades the role version to V7 if
+// the client version passed through the gRPC metadata is below the version
+// specified in minSupportedRoleV8Version.
+func maybeDowngradeRoleVersionToV7(role *types.RoleV6, clientVersion *semver.Version) *types.RoleV6 {
```

Comment:
shouldn't you modify the role so that it won't grant access to the saml IDP apps if the v8 role wouldn't grant access?